### PR TITLE
feat: add getIgnoredData() to the ethereum storage

### PR DESCRIPTION
**Pull request by vrolland**
## Description of the changes

Add an entry to retry to get previous ignored data

## Next step

Call this entry point from the data-access

### CODE REVIEW

File: packages/ethereum-storage/src/ethereum-entries-to-ipfs-content.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ const SAFE_MAX_HEADER_SIZE: number = 500;
  * @param ethereumEntries Ethereum entries from the smart contract
  * @returns Filtered list of dataId with metadata
  */
-export default async function ethereumEntriesToEntries(
+export default async function œ(
```
Comment (benjlevesque) [inline packages/ethereum-storage/src/ethereum-entries-to-ipfs-content.ts:24]: I don't think this is what you meant 😁

File: packages/ethereum-storage/src/ethereum-storage.ts [hunk 2]
```diff
@@ -391,6 +391,53 @@ export default class EthereumStorage implements StorageTypes.IStorage {
...
+      this.ipfsManager,
+      this.ignoredDataIds,
+      this.logger,
+      this.maxConcurrency,
+    );
+
+    const ids = entries.map(entry => entry.id) || [];
+    // Pin data asynchronously
+    // tslint:disable-next-line:no-floating-promises
+    this.pinDataToIPFS(ids);
```
Comment (benjlevesque) [inline packages/ethereum-storage/src/ethereum-storage.ts:427]: why asynchronously? and what happens if it fails?

File: packages/ethereum-storage/src/ignored-dataIds.ts [hunk 3]
```diff
@@ -121,6 +146,17 @@ export default class IgnoredDataIds {
     return result;
   }
 
+  /**
+   * Check if it is the time to retry the entry
+   * @param entry to check
+   * @returns true if it is time to retry
+   */
+  private shouldBeRetry(
```
Comment (benjlevesque) [inline packages/ethereum-storage/src/ignored-dataIds.ts:154]:
```suggestion
  private shouldRetry(
```

File: packages/ethereum-storage/src/ignored-dataIds.ts [hunk 4]
```diff
@@ -121,6 +146,17 @@ export default class IgnoredDataIds {
...
+    entry: StorageTypes.IIgnoredDataId,
+  ): boolean {
+    return entry.toRetry && (entry.lastTryTimestamp as number + Math.floor(Math.exp(entry.iteration)) * INTERVAL_RETRY_MS) <= Date.now();
```
Comment (benjlevesque) [inline packages/ethereum-storage/src/ignored-dataIds.ts:157]: you should explain this calculation in plain english

File: packages/ethereum-storage/test/ethereum-storage.test.ts [hunk 5]
```diff
@@ -809,6 +809,68 @@ describe('EthereumStorage', () => {
...
+      ethereumStorage.ipfsManager.read = chai.spy(
+        async (_hash: string): Promise<StorageTypes.IIpfsObject> => ({
+          content: 'ok',
+          ipfsLinks: [],
+          ipfsSize: 2,
+        }),
+      );
+
+      const entries = await ethereumStorage.getIgnoredData();
+      expect(entries.length, 'config wrong').to.be.equal(1);
```
Comment (benjlevesque) [inline packages/ethereum-storage/test/ethereum-storage.test.ts:848]:
```suggestion
      expect(entries.length, 'config wrong').to.equal(1);
```

File: packages/ethereum-storage/test/ethereum-storage.test.ts [hunk 6]
```diff
@@ -809,6 +809,68 @@ describe('EthereumStorage', () => {
...
+      expect(entries[0], 'config wrong').to.be.deep.equal({
```
Comment (benjlevesque) [inline packages/ethereum-storage/test/ethereum-storage.test.ts:849]:
```suggestion
      expect(entries[0], 'config wrong').to.deep.equal({
```

**Comment by benjlevesque** [comment]
It's alarming that tests passed with the typo in `packages/ethereum-storage/src/ethereum-entries-to-ipfs-content.ts`

File: packages/ethereum-storage/src/ethereum-entries-to-ipfs-content.ts (hunk 1 above)
Comment (vrolland) [reply packages/ethereum-storage/src/ethereum-entries-to-ipfs-content.ts:24]: interesting typo !

File: packages/ethereum-storage/src/ethereum-storage.ts (hunk 2 above)
Comment (vrolland) [reply packages/ethereum-storage/src/ethereum-storage.ts:427]: Because the pining can be done afterward, If it fails it will just be warning in the logger.

Comment (benjlevesque) [reply packages/ethereum-storage/src/ethereum-storage.ts:427]: then we should track the warnings in the logger I guess ;) there are ways to create metrics or alerts for that in GCP logs

Comment (vrolland) [reply packages/ethereum-storage/src/ethereum-storage.ts:427]: Yes, it would a very nice to have! I Jira it !
